File: leetcode/p1396/Solution.java
class UndergroundSystem {

    static class StartEnd {
        String start;
        String end;
        public StartEnd(String start, String end) {
            this.start = start;
            this.end = end;
        }
        @Override
        public int hashCode() {
            return Objects.hash(start, end);
        }
        
        @Override
        public boolean equals(Object o) {
            if (!(o instanceof StartEnd)) {
                return false;
            }
            StartEnd other = (StartEnd)o;
            return start.equals(other.start) && end.equals(other.end);
        }
    }

    Map<Integer, Integer> startTimeMap = new HashMap<>();
    Map<Integer, String> startStationMap = new HashMap<>();
    Map<StartEnd, Integer> totalMap = new HashMap<>();
    Map<StartEnd, Integer> countMap = new HashMap<>();
    
    public UndergroundSystem() {
        
    }
    
    public void checkIn(int id, String stationName, int t) {
        startTimeMap.put(id, t);
        startStationMap.put(id, stationName);
    }
    
    public void checkOut(int id, String stationName, int t) {
        int startTime = startTimeMap.get(id);
        String startStation = startStationMap.get(id);
        StartEnd key = new StartEnd(startStation, stationName);
        totalMap.put(key, totalMap.getOrDefault(key,0)+t - startTime);
        countMap.put(key, countMap.getOrDefault(key,0)+1);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        StartEnd key = new StartEnd(startStation, endStation);
        return (double)totalMap.get(key) / countMap.get(key);
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: leetcode/p1396/Solution.swift
class UndergroundSystem {
    private var inMap = [Int : (String, Int)] ()
    private var timeMap = [String : [String : (Int, Int)]]()

    init() {
        
    }
    
    func checkIn(_ id: Int, _ stationName: String, _ t: Int) {
        inMap[id] = (stationName, t)
    }
    
    func checkOut(_ id: Int, _ stationName: String, _ t: Int) {
        guard let (startStation, startTime) = inMap[id] else {
            return
        }

        var map = timeMap[startStation, default: [String : (Int, Int)]()]
        let (sum, count) = map[stationName, default: (0, 0)]
        map[stationName] = (sum + (t - startTime), count + 1)
        timeMap[startStation] = map
    }
    
    func getAverageTime(_ startStation: String, _ endStation: String) -> Double {
        guard let map = timeMap[startStation] else {
            return 0.0
        }

        guard let (sum, count) = map[endStation] else {
            return 0.0
        }

        return Double(sum) / Double(count)
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * let obj = UndergroundSystem()
 * obj.checkIn(id, stationName, t)
 * obj.checkOut(id, stationName, t)
 * let ret_3: Double = obj.getAverageTime(startStation, endStation)
 */
